protocols/raft/src/main/java/io/atomix/protocols/raft/roles/FollowerRole.java

```diff
@@ -251,49 +233,109 @@ private void resetHeartbeatTimeoutFromDifferentThread() {
     raft.runOnHeartbeatContext(this::resetHeartbeatTimeoutFromHeartbeatThread);
   }
 
-  /** Resets the heartbeat timer. */
   private void resetHeartbeatTimeoutFromHeartbeatThread() {
     raft.checkHeartbeatThread();
     if (!isRunning()) {
       return;
     }
 
-    final long currentHeartbeatTime = System.currentTimeMillis();
-    if (lastHeartbeat > 0) {
-      raft.getRaftRoleMetrics().observeHeartbeatInterval(currentHeartbeatTime - lastHeartbeat);
-    }
-    lastHeartbeat = currentHeartbeatTime;
-
-    // If a timer is already set, cancel the timer.
-    if (heartbeatTimoutTimer != null) {
-      heartbeatTimoutTimer.cancel();
-    }
+    cancelHeartbeatTimer();
+    updateHeartbeat(System.currentTimeMillis());
 
     // Set the election timeout in a semi-random fashion with the random range
     // being election timeout and 2 * election timeout.
     final Duration delay =
         raft.getElectionTimeout()
             .plus(Duration.ofMillis(random.nextInt((int) raft.getElectionTimeout().toMillis())));
-    heartbeatTimoutTimer =
-        raft.getThreadContext()
+    heartbeatTimer = raft.getHeartbeatThread().schedule(delay, () -> onHeartbeatTimeout(delay));
+  }
+
+  private void updateHeartbeat(final long currentTimestamp) {
+    raft.checkHeartbeatThread();
+
+    if (lastHeartbeat > 0) {
+      raft.getRaftRoleMetrics().observeHeartbeatInterval(currentTimestamp - lastHeartbeat);
+    }
+
+    lastHeartbeat = currentTimestamp;
+  }
+
+  private void schedulePollRequests(final Duration delay) {
+    raft.checkThread();
+
+    if (!isRunning()) {
+      return;
+    }
+
+    if (raft.getFirstCommitIndex() == 0 || raft.getState() == RaftContext.State.READY) {
+      final long missTime = System.currentTimeMillis() - lastHeartbeat;
+      log.info(
```

Comment:
It was already there; it was useful for me while trying to figure this out, and it can help explain why some elections are triggered. It shouldn't happen too often anyway because this only happens when no heartbeats are received.

Reply:
:+1: